keanu-project/src/main/java/io/improbable/keanu/vertices/VertexLabel.java

```diff
@@ -1,30 +1,58 @@
 package io.improbable.keanu.vertices;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 import java.util.Objects;
 
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableList;
+
 public class VertexLabel {
-    private final String label;
+    private static final char NAMESPACE_SEPARATOR = '.';
+    private static final char NAME_PREPENDER = ':';
+    private final String name;
+    private final List<String> namespace;
+
+    public VertexLabel(String name, String... namespace) {
+        this.name = name;
+        this.namespace = ImmutableList.copyOf(namespace);
+    }
+
+    public VertexLabel inNamespace(String topLevelNamespace) {
+        List<String> newNamespace = ImmutableList.<String>builder().addAll(namespace).add(topLevelNamespace).build();
+        return new VertexLabel(this.name, newNamespace.toArray(new String[0]));
+    }
 
-    public VertexLabel(String label) {
-        this.label = label;
+    public VertexLabel removeOuterNamespace() throws VertexLabelException {
+        try {
+            List<String> reducedNamespace = namespace.subList(0, namespace.size()-1);
+            return new VertexLabel(this.name, reducedNamespace.toArray(new String[0]));
+        } catch (IndexOutOfBoundsException e) {
+            throw new VertexLabelException("There is no namespace to remove", e);
+        }
     }
 
     @Override
     public String toString() {
-        return label;
+        ArrayList reversedNamespace = new ArrayList(namespace);
```

Comment:
Could we use Lists.reverse() here to avoid having to create a copy and having to reverse in place?

Reply:
much better - thanks.